zilencer/views.py

```diff
@@ -96,3 +112,69 @@ def remote_server_notify_push(request: HttpRequest, entity: Union[UserProfile, R
+def add_payment_method(request: HttpRequest) -> HttpResponse:
+    user = request.user
+    ctx = {
+        "publishable_key": STRIPE_PUBLISHABLE_KEY,
+        "email": user.email,
+    }
+    if not user.is_realm_admin:
+        ctx["error_message"] = _("You should be an administrator of the organization %s to view this page."
+                                 % (user.realm.name,))
+        return render(request, 'zilencer/payment.html', context=ctx)
+
+    try:
+        if request.method == "GET":
+            try:
+                customer_obj = Customer.objects.get(realm=user.realm)
+                cards = stripe.Customer.retrieve(customer_obj.stripe_customer_id).sources.all(object="card")
+                ctx["num_cards"] = len(cards["data"])
+            except Customer.DoesNotExist:
+                ctx["num_cards"] = 0
+            return render(request, 'zilencer/payment.html', context=ctx)
+
+        if request.method == "POST":
+            token = request.POST.get("stripeToken", "")
+            # The card metadata doesn't show up in Dashboard but can be accessed
+            # using the API.
+            card_metadata = {"added_user_id": user.id, "added_user_email": user.email}
+            try:
+                customer_obj = Customer.objects.get(realm=user.realm)
+                customer = stripe.Customer.retrieve(customer_obj.stripe_customer_id)
+                customer.sources.create(source=token, metadata=card_metadata)
+                ctx["num_cards"] = len(customer.sources.all(object="card")["data"])
+            except Customer.DoesNotExist:
+                customer_metadata = {"string_id": user.realm.string_id}
+                # Description makes it easier to identify customers in Stripe dashboard
+                description = "{} ({})".format(user.realm.name, user.realm.string_id)
+                customer = stripe.Customer.create(source=token,
+                                                  description=description,
+                                                  metadata=customer_metadata)
+
+                card = customer.sources.all(object="card")["data"][0]
+                card.metadata = card_metadata
+                card.save()
+                Customer.objects.create(realm=user.realm, stripe_customer_id=customer.id)
+                ctx["num_cards"] = 1
+            ctx["payment_method_added"] = True
+            return render(request, 'zilencer/payment.html', context=ctx)
+    except (CardError, RateLimitError, APIConnectionError) as e:
+        err = e.json_body.get('error', {})
+        logging.error("Stripe error - Status: {}, Type: {}, Code: {}, Param: {}, Message: {}".format(
+            e.http_status, err.get('type'), err.get('code'), err.get('param'), err.get('message')
+        ))
+        ctx["error_message"] = err.get('message')
+        return render(request, 'zilencer/payment.html', context=ctx)
+    except (InvalidRequestError, AuthenticationError, StripeError) as e:
```

Comment:
(done)

Reply:
Ahh. I thought these are separate category of errors after reading this. 

> These are Stripe-specific errors that don't fall into any of the above categories.  You'll want to log these, send a notification, and inform your user that something went wrong.

https://groups.google.com/a/lists.stripe.com/forum/#!topic/api-discuss/frOL8pBxaas

Reply:
Interesting. Yeah, I think the author of that email was speaking imprecisely (possibly understood it imprecisely) -- they really meant that a StripeError is a Stripe-specific error, and if you catch one after you already have handlers for the others, it doesn't fall into any of the above categories.

(That email is also discussing the Ruby client library, so it's possible that works differently, though I wouldn't expect so.)

This relationship isn't clearly documented anywhere that I can see. The way I found it is:
* I figured it really ought to be this way, because that's the only reasonable way to write the API and surely the Stripe folks were competent enough to get this right.
* Then I double-checked by reading the source code. :-) `/srv/zulip-py3-venv/lib/.../stripe/error.py` in my dev environment. Fortunately the code is quite short and clear.